be2-scala/src/main/scala/ch/epfl/pop/pubsub/graph/handlers/ElectionHandler.scala

```diff
@@ -59,27 +62,80 @@ class ElectionHandler(dbRef: => AskableActorRef) extends MessageHandler {
     Await.result(ask, duration)
   }
 
-   def handleResultElection(rpcMessage: JsonRpcRequest): GraphMessage = Right(
+  def handleResultElection(rpcMessage: JsonRpcRequest): GraphMessage = Right(
     PipelineError(ErrorCodes.SERVER_ERROR.id, "NOT IMPLEMENTED: ElectionHandler cannot handle ResultElection messages yet", rpcMessage.id)
   )
 
-  def handleEndElection(rpcMessage: JsonRpcRequest): GraphMessage = Right(
-    PipelineError(ErrorCodes.SERVER_ERROR.id, "NOT IMPLEMENTED: ElectionHandler cannot handle EndElection messages yet", rpcMessage.id)
-  )
+  def handleEndElection(rpcMessage: JsonRpcRequest): GraphMessage = {
+    val witnessSignatures = rpcMessage.getParamsMessage match {
+      case Some(it) => it.witness_signatures.map(_.signature)
+      case _ => Nil
+    }
+    val electionChannel: Channel = rpcMessage.getParamsChannel
+    val combined = for {
+      electionQuestionResults <- createElectionQuestionResults(electionChannel)
+      // propagate the endElection message
+      _ <- dbAskWritePropagate(rpcMessage)
+      //data to be broadcast
+      resultElection: ResultElection = ResultElection(electionQuestionResults, witnessSignatures)
+      data: Base64Data = Base64Data.encode(resultElectionFormat.write(resultElection).toString)
+      // create & propagate the resultMessage
+      _ <- dbBroadcast(rpcMessage, electionChannel, data, electionChannel)
+    } yield ()
+    Await.ready(combined, duration).value match {
+      case Some(Success(_)) => Left(rpcMessage)
+      case _ => Right(PipelineError(ErrorCodes.SERVER_ERROR.id, s"handleEndElection unknown error", rpcMessage.getId))
+    }
+  }
+
+  private def createElectionQuestionResults(electionChannel: Channel): Future[List[ElectionQuestionResult]] = {
+    for {
+      votesList <- electionChannel.getLastVotes(dbActor)
+      castsVotesElections = votesList.map(_._2)
+      setupMessage <- electionChannel.getSetupMessage(dbActor)
+      questionToBallots = setupMessage.questions.map(question => question.id -> question.ballot_options).toMap
+      resultsTable = mutable.HashMap.from(for {
+        (question, ballots) <- questionToBallots
+      } yield question -> ballots.map(_ -> 0).toMap)
+      castVoteElection <- Future.traverse(castsVotesElections) { castVoteElection => {
+        for {
+          voteElection <- castVoteElection.votes
+          _ = voteElection.vote match {
+            case Some(List(index)) =>
+              val question = voteElection.question
+              val ballots = questionToBallots(question).toArray
+              val ballot = ballots.apply(index)
+              val questionResult = resultsTable(question)
+              resultsTable.update(question, questionResult.updated(ballot, questionResult(ballot) + 1))
+            case _ =>
+          }
+        } {}
+        Future(Success(()))
+      }
+      }
+      results = resultsTable.map(tuple => {
+        val (qid, ballotToCount) = tuple
```

Comment:
In 52b4cbd, I pointed out that there's an elegant way to avoid binding a name there. You can also write the lambda as `{ case (qid, ballotToCount) =>`, that works too